src/circuit/gadget/sinsemilla/chip/hash_to_point.rs

```diff
@@ -0,0 +1,446 @@
+            let mut zs = Vec::with_capacity(piece.num_words() + 1);
+
+            // Copy message and initialize running sum `z` to decompose message in-circuit
+            let cell = region.assign_advice(
+                || "z_0 (copy of message)",
+                config.bits,
+                offset,
+                || piece.field_elem().ok_or(Error::SynthesisError),
+            )?;
+            region.constrain_equal(&config.perm, piece.cell(), cell)?;
+            zs.push(CellValue::new(cell, piece.field_elem()));
+
+            // Assign cumulative sum such that
+            //          z_i = 2^K * z_{i + 1} + m_{i + 1}
+            // => z_{i + 1} = (z_i - m_{i + 1}) / 2^K
+            //
+            // For a message m = m_1 + 2^K m_2 + ... + 2^{K(n-1)} m_n}, initialize z_0 = m.
+            // We end up with z_n = 0.
+            let mut z = piece.field_elem();
+            let inv_2_k = pallas::Base::from_bytes(&INV_TWO_POW_K).unwrap();
+
+            // We do not assign the final z_n as it is constrained to be zero.
+            for (idx, word) in words[0..(words.len() - 1)].iter().enumerate() {
+                // z_{i + 1} = (z_i - m_{i + 1}) / 2^K
+                z = z
+                    .zip(*word)
+                    .map(|(z, word)| (z - pallas::Base::from_u64(word as u64)) * inv_2_k);
+                let cell = region.assign_advice(
+                    || format!("z_{:?}", idx + 1),
+                    config.bits,
+                    offset + idx + 1,
+                    || z.ok_or(Error::SynthesisError),
+                )?;
+                zs.push(CellValue::new(cell, z))
+            }
+
+            zs
+        };
+
+        // Copy in the accumulator x-coordinate
+        let mut x_a: X<pallas::Base> = copy(
+            region,
+            || "Initialize accumulator x-coordinate",
+            config.x_a,
+            offset,
+            &x_a.0,
+            &config.perm,
+        )?
+        .into();
+
+        let mut y_a = y_a;
+
+        let generators: Vec<Option<(pallas::Base, pallas::Base)>> =
+            if let Some(generators) = generators {
+                generators.into_iter().map(Some).collect()
+            } else {
+                vec![None; piece.num_words()]
+            };
+
+        for (row, gen) in generators.iter().enumerate() {
+            let x_p = gen.map(|gen| gen.0);
+            let y_p = gen.map(|gen| gen.1);
+
+            // Assign `x_p`
+            region.assign_advice(
+                || "x_p",
+                config.x_p,
+                offset + row,
+                || x_p.ok_or(Error::SynthesisError),
+            )?;
+
+            // Compute and assign `lambda_1`
+            let lambda_1 = {
+                let lambda_1 = x_a
+                    .value()
+                    .zip(y_a.0)
+                    .zip(x_p)
+                    .zip(y_p)
+                    .map(|(((x_a, y_a), x_p), y_p)| (y_a - y_p) * (x_a - x_p).invert().unwrap());
```

Comment:
Now that https://github.com/zcash/halo2/pull/293 has merged, we can batch these inversions.

Do not make this change in this PR; I want to do this as a separate optimization (I'm just noting it here so we remember).